### PR TITLE
circt: init at 1.29.0

### PR DESCRIPTION
**Pull request by SharzyL**
Circuit IR Compilers and Tools (https://github.com/llvm/circt)

Pending issues:
- I am experiencing sporadic internal compiler error when building circt on both my pc and server. The error remains occuring even when switching from `stdenv` to `clangStdenv`, but disappears when building with `--cores 1`. May somebody help testing on his/her machine (`nix build github:SharzyL/nixpkgs/circt#circt`)?

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/compilers/circt/default.nix [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+  pname = "circt";
+  version = "1.29.0";
+  src = fetchFromGitHub {
+    owner = "llvm";
+    repo = "circt";
+    rev = "firtool-${version}";
+    sha256 = "sha256-HsXwh98RZuXvK/KkZ2NAGwWNLxUAQVj+WKzZXd4C4Is=";
+    fetchSubmodules = true;
+  };
+  psutil = python3.withPackages (ps: [ ps.psutil ]);
```
Comment (NickCao) [inline pkgs/development/compilers/circt/default.nix:21]: the naming convention is `pythonEnv`

File: pkgs/development/compilers/circt/default.nix [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+let
+  pname = "circt";
+  version = "1.29.0";
+  src = fetchFromGitHub {
+    owner = "llvm";
+    repo = "circt";
+    rev = "firtool-${version}";
+    sha256 = "sha256-HsXwh98RZuXvK/KkZ2NAGwWNLxUAQVj+WKzZXd4C4Is=";
+    fetchSubmodules = true;
+  };
```
Comment (NickCao) [inline pkgs/development/compilers/circt/default.nix:20]: Just use `rec`, let in is hard to read.

File: pkgs/development/compilers/circt/default.nix [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+      "-DLLVM_BUILD_EXAMPLES=OFF"
+      "-DLLVM_OPTIMIZED_TABLEGEN=ON"
+      "-DLLVM_ENABLE_PROJECTS=mlir"
+      "-DLLVM_EXTERNAL_PROJECTS=circt"
+      "-DLLVM_EXTERNAL_CIRCT_SOURCE_DIR=.."
+      "-DCIRCT_LLHD_SIM_ENABLED=OFF"
+    ];
+
+    preConfigure = ''
+      substituteInPlace test/circt-reduce/test/annotation-remover.mlir --replace "/usr/bin/env" "${coreutils}/bin/env"
```
Comment (NickCao) [inline pkgs/development/compilers/circt/default.nix:40]: use `patchShebangs`

File: pkgs/development/compilers/circt/default.nix [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+    '';
+
+    installPhase = ''
+      mkdir -p $out/bin
+      mv bin/{{fir,hls}tool,circt-{as,dis,lsp-server,opt,reduce,translate}} $out/bin
```
Comment (NickCao) [inline pkgs/development/compilers/circt/default.nix:45]: don't forget to add `runHook {pre,post}Install`

File: pkgs/development/compilers/circt/default.nix (hunk 3 above)
Comment (SharzyL) [reply pkgs/development/compilers/circt/default.nix:39]: This is not a shebang, but a [hardcoded path in test spec](https://github.com/llvm/circt/blob/main/test/circt-reduce/test/annotation-remover.mlir#L3). `patchShebangs` seems not working for this?

**Review by NickCao — COMMENTED** [review]
looks mostly ok, just fix the indentation with nixpkgs-fmt, and figure out what to do with the ICE.

**Comment by SharzyL** [comment]
`nixpkgs-fmt` is done.

It seems that ICE is caused by hardware issue. It disappears when I lower the overclocking configuration. Now it works fine.

**Comment by NickCao** [comment]
And you may with to schedule it as big-parallel:
```
requiredSystemFeatures = [ "big-parallel" ];
```

**Comment by SharzyL** [comment]
`big-parallel` done

**Review by NickCao — APPROVED** [review]
diff LGTM, build running.

**Comment by NickCao** [comment]
Result of `nixpkgs-review pr 214870` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>circt</li>
  </ul>
</details>

**Comment by dtzSiFive** [comment]
FWIW: https://github.com/dtzSiFive/circt-nix (updated regularly to latest git, backed by cachix).  LLVM is split out so it's not rebuilt on every update and can be built in somewhat smaller pieces (although we bump LLVM weekly so that mostly helps folks using this very actively / cachix-friendliness).  Includes development shell for anyone hacking on CIRCT.

When I get some cycles, I'll try to revive the MLIR PR to nixpkgs and help here as needed :+1:, not sure when that'll be :sweat_smile:.  FWIW CIRCT makes releases roughly weekly now.

**Pull request merged**